File: Java/src/main/java/org/chess/Engine.java
package org.chess;

import org.chess.domain.GameState;
import org.chess.rendering.ConsoleRenderer;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Engine {

    private final ConsoleRenderer renderer = new ConsoleRenderer();

    private boolean running = false;

    public void start() {
        running = true;
        var initialState = GameState.initialState();
        var consoleReader = new BufferedReader(new InputStreamReader(System.in));
        gameLoop(initialState, consoleReader);
    }

    private void gameLoop(GameState initialState, BufferedReader reader) {
        GameState gameState = initialState;
        while (running) {
            clearScreen();
            renderToScreen(gameState);
            promptPlayerInput();
            final String playerInput = readPlayerInput(reader);
            gameState = handlePlayerInput(playerInput, gameState);
        }
    }

    private void clearScreen() {
        System.out.print("\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n");
    }

    private void renderToScreen(GameState gameState) {
        System.out.println(renderer.render(gameState));
    }

    private void promptPlayerInput() {
        System.out.println("Type your move in the format \"from -> to\", e.g., \"e2 -> e8\".");
    }

    private String readPlayerInput(BufferedReader reader) {
        try {
            return reader.readLine();
        } catch (IOException e) {
            System.out.println("Failed to read player input. Exiting game.");
            e.printStackTrace();
            running = false;
            return "";
        }
    }

    private GameState handlePlayerInput(String playerInput, GameState gameState) {
        // For now, we do not transform the game state. The game state will remain the same for
        // any player input. Later we will parse the player input and transform the game state
        // according to the rules of chess.
        return gameState;
    }
}

File: Kotlin/src/main/kotlin/org/chess/Engine.kt
package org.chess

import org.chess.domain.GameState
import org.chess.domain.PlayerColor
import org.chess.input.MoveCommandParser
import org.chess.rendering.ConsoleRenderer
import org.chess.ruleengine.RuleEngine
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class Engine {

    private val renderer = ConsoleRenderer()
    private val moveCommandParser = MoveCommandParser()
    private val ruleEngine = RuleEngine()

    private var running = false

    fun start() {
        running = true
        val initialState = GameState.initialState()
        val consoleReader = BufferedReader(InputStreamReader(System.`in`))
        gameLoop(initialState, consoleReader)
    }

    private fun gameLoop(initialState: GameState, reader: BufferedReader) {
        var gameState = initialState
        while (running) {
            renderToScreen(gameState)
            promptPlayerInput(gameState)
            val playerInput = readPlayerInput(reader)
            gameState = handlePlayerInput(playerInput, gameState)
        }
    }

    private fun renderToScreen(gameState: GameState) {
        println(renderer.render(gameState))
    }

    private fun promptPlayerInput(gameState: GameState) {
        val playerName = toPlayerName(gameState.currentPlayer)
        System.out.printf("%s's turn.\n", playerName)
        println("Type your move in the format \"from -> to\", e.g., \"e2 -> e8\".")
    }

    private fun toPlayerName(playerColor: PlayerColor): String {
        return when (playerColor) {
            PlayerColor.WHITE -> "White"
            PlayerColor.BLACK -> "Black"
        }
    }

    private fun readPlayerInput(reader: BufferedReader): String {
        return try {
            reader.readLine()
        } catch (e: IOException) {
            println("Failed to read player input. Exiting game.")
            e.printStackTrace()
            running = false
            ""
        }
    }

    private fun handlePlayerInput(playerInput: String, gameState: GameState): GameState {
        val move = try {
            moveCommandParser.parse(playerInput)
        } catch (e: IllegalArgumentException) {
            println("Please make sure your command is in the correct format.")
            return gameState
        }

        val moveResult = ruleEngine.execute(move, gameState)
        if (!moveResult.violations.isEmpty()) {
            println("Invalid move!")
            return gameState
        }

        return moveResult.gameState
    }
}
